static/js/typing.js

```diff
@@ -0,0 +1,107 @@
+            send_typing_notification_ajax(compose.recipient(), operation);
+        }
+    } else if (operation === 'stop') {
+        if (compose.recipient()) {
+            send_typing_notification_ajax(compose.recipient(), operation);
+        }
+    }
+}
+
+function check_and_send_start() {
+    check_and_send('start');
+}
+
+function get_throttled_function() {
+    return _.throttle(check_and_send_start,
+        TYPING_STARTED_SEND_FREQUENCY,
+        {trailing: false});
+}
+
+function check_and_send_stop() {
+    check_and_send('stop');
+    exports.send_start_notification = get_throttled_function();
+    $(document).on('input', '#new_message_content', exports.send_start_notification);
+}
+
+exports.send_start_notification = get_throttled_function();
+exports.send_stop_notification = _.debounce(check_and_send_stop, TYPING_STOPPED_WAIT_PERIOD);
+
+$(document).on('compose_canceled.zulip', check_and_send_stop);
+$(document).on('compose_finished.zulip', check_and_send_stop);
+$(document).on('blur', '#new_message_content', check_and_send_stop);
+$(document).on('input', '#new_message_content', exports.send_start_notification);
+$(document).on('input', '#new_message_content', exports.send_stop_notification);
+
+function full_name(email) {
+    return people.get_by_email(email).full_name;
+}
+
+exports.hide_notification = function (event) {
+    // The typing notification is also sent to the user who is typing
+    // In this case the notification is not displayed
+    if (event.sender.user_id !== page_params.user_id) {
+        $('#typing_notifications').find('[data-email="' + event.sender.email + '"]').remove();
+        if (users_currently_typing.length === 1) {
+            $('#typing_notifications').hide();
+        }
+        var i = users_currently_typing.indexOf(event.sender.user_id);
+        users_currently_typing.splice(i);
+    }
+};
+
+var debounced_hide_notification = _.debounce(exports.hide_notification,
+  TYPING_STARTED_EXPIRY_PERIOD);
+
+exports.display_notification = function (event) {
+    // The typing notification is also sent to the user who is typing
+    // In this case the notification is not displayed
+    if (event.sender.user_id !== page_params.user_id) {
+        var recipientsMatchingID = event.recipients.filter(function (recipient) {
+            return recipient.user_id === page_params.user_id;
+        });
+        var userIsRecipient = recipientsMatchingID.length > 0;
+        event.sender.name = full_name(event.sender.email);
+        if (narrow.narrowed_to_pms() && userIsRecipient) {
+            $('#typing_notifications').show();
+            if ($('#typing_notifications').find('[data-email="' + event.sender.email + '"]').length === 0) {
+                $('#typing_notifications ul').append(templates.render('typing_notification', event.sender));
+                users_currently_typing.push(event.sender.user_id);
+            }
+            debounced_hide_notification(event);
```

Comment:
Fixed the names!

## New datastructure
The `users_currently_typing` data structure now looks like `{group: users_typing}` where `group` is a list of user ids in the  conversation and `users_typing` is a list of user ids for users currently typing in this conversation.

## Updating the datastructure
When a typing event is received, the sender is added to (or removed from) `users_currently_typing[recipients]` where `recipients` is a list of user ids from the event's recipients. The sender is in this list as well.

## Rendering the notifications
For pm-with narrows, the group is the current_user + users from the operand. `users_currently_typing` is looked up and the list is used to render the typing_notifications template.

For the private messages narrow, all users currently typing (in all conversations that the current user is a recipient of) are displayed.